### PR TITLE
generate-repology-metadata: add urls to patches and build recipe to generated json

### PR DESCRIPTION
**Pull request by Grimler91**
And add helper functions for writing json objects, to make it easier to extend and maintain the script.

Fixes https://github.com/termux/repology-metadata/issues/2.

Example packages.json: https://grimler.se/files/packages.json

### CODE REVIEW

**Comment by Grimler91** [comment]
Ping @AMDmi3, do you have any comments on the entries and format?

File: generate-repology-metadata.sh [hunk 1]
```diff
@@ -24,41 +25,97 @@ export TERMUX_ARCH=aarch64
...
 	if [ "$TERMUX_PKG_SRCURL" != "" ]; then
-		echo "    \"srcurl\": \"$TERMUX_PKG_SRCURL\","
+		print_json_element "srcurl" "$TERMUX_PKG_SRCURL"
 	fi
+	print_json_element "maintainer" "$TERMUX_PKG_MAINTAINER"
+
+	print_json_element "package_sources_url" "${repo_url}/tree/master/$(dirname $(git ls-files --full-name build.sh))"
+	print_json_element "package_recipe_url" "${repo_url}/blob/master/$(git ls-files --full-name build.sh)"
+	print_json_element "package_recipe_url_raw" "${repo_raw_url}/master/$(git ls-files --full-name build.sh)"
+	local patches=$(git ls-files --full-name "*.patch" "*.patch32" "*.patch64" "*.patch.beforehostbuild")
```
Comment (thunder-coding) [inline generate-repology-metadata.sh:111]:
```suggestion
	local patches=$(git ls-files --full-name "*.patch" "*.patch32" "*.patch64" "*.patch.beforehostbuild" "*.diff" "*.patch.txt")
```

File: generate-repology-metadata.sh [hunk 2]
```diff
@@ -24,41 +25,97 @@ export TERMUX_ARCH=aarch64
...
 	local pkg=$(basename $path)
+	pushd $path > /dev/null
+	repo_url="$(git config --get remote.origin.url)"
+	# Convert to normal https url if it starts with git@
+	if [ "${repo_url:0:4}" == "git@" ]; then
+		repo_url="$(echo $repo_url|sed  -e 's%:%/%g' -e 's%git@%https://%g' )"
+	fi
+	# To get raw url we assume that repositories are hosted at github
+	repo_raw_url="$(echo $repo_url|sed 's%https://github.com%https://raw.githubusercontent.com%g')"
+
```
Comment (thunder-coding) [inline generate-repology-metadata.sh:90]: Would make sense to add a condition to confirm whether it is really a valid GitHub repo URL.

**Review by thunder-coding — CHANGES_REQUESTED** [review]

File: generate-repology-metadata.sh (hunk 1 above)
Comment (Grimler91) [reply generate-repology-metadata.sh:111]: Thanks for the review! Will fix, will rename all .patch.txt to .diff, as having different suffix'es seem unnecessary

File: generate-repology-metadata.sh (hunk 2 above)
Comment (Grimler91) [reply generate-repology-metadata.sh:90]: Sure, can match a regex pattern instead and error if not match!

**Comment by Grimler91** [comment]
Accidentally pushed to master, but comments have been addressed, url is verified that it is a github url, and .diff are included in patch list

**Comment by Grimler91** [comment]
Hmm, no, I pushed the wrong branch to master. Will have to force push to fix this

**Pull request closed without merging**